### ISSUE TITLE
.NET Native: Wrong result from LINQ query with condition based on navigation property

### ISSUE THREAD

**Issue by hameedfs** [dotnet/efcore#5841]
I am using EFCore in-memory provider in UWP application and compiling in Release mode with "Compile with .Net Native tool chain" option enabled.

`dbContext.WorkItems.Count(w => w.Employee.Team.Name == <any value>)` always returns zero with the following entity definitions.

```
public class Team
{
    public int TeamId { get; set; }
    public string Name { get; set; }
    public ICollection<Employee> Employees { get; set; }
}

public class Employee
{
    public int EmployeeId { get; set; }
    public string Name { get; set; }
    public int? TeamId { get; set; }
    public Team Team { get; set; }
    public ICollection<WorkItem> WorkItems { get; set; }
}

public class WorkItem
{
    public int WorkItemId { get; set; }
    public string Name { get; set; }
    public int EmployeeId { get; set; }
    public Employee Employee { get; set; }
}
```

Note that it works fine if any one of the following is done.
- "Compile with .Net Native tool chain" option is disabled.
- TeamId property in Employee entity is made not-nullable.
- Sqlite database provider is used instead of in-memory.

I am using Visual Studio 2015 Update 2 along with other latest updates.
EFCore NuGet package: Microsoft.EntityFrameworkCore.InMemory version 1.0.0-rc2-final
OS: Windows 10

The repro solution is attached. Debug-x86 returns the correct count while Release-x86 returns zero.

[EFNativeToolChain.zip](https://github.com/aspnet/EntityFramework/files/329378/EFNativeToolChain.zip)

**Comment by rowanmiller** [comment]
Sounds like it may be a .NET Native issue, we should isolate it down and hand off if needed.

**Comment by natemcmaster** [comment]
I confirmed this repros with VS 2015 U3 and 1.0.0. I haven't yet identified the cause of the issue. I'm going to continue investigating.

**Comment by natemcmaster** [comment]
From my investigation, it appears EF Core is working as expected. I've passed off the query bug to the .NET Native team for more help debugging a compiled lambda expression.

For reference, here is the lambda expression EF generates and executes for the in-memory provider:
[expression.txt](https://github.com/aspnet/EntityFramework/files/394975/expression.txt)
[log.txt](https://github.com/aspnet/EntityFramework/files/394976/expressionprinter.txt)


@hameedfs just FYI, the reason SQLite works in this case is that it actually pulls all entities in memory and the uses LINQ to Objects to translate the rest of the query. The log contains this warning:

> Warning: The LINQ expression 'DefaultIfEmpty()' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.

**Comment by MattWhilden** [comment]
@natemcmaster can we sync on this tomorrow? I want to make sure this is properly tracked for the next update of .NET Native.

**Comment by AndriySvyryd** [comment]
Verified that this is fixed in the prerelease builds of EF 2.0.1 and UWP 6.0.0
